packages/muelu/src/Graph/Containers/MueLu_Aggregates_kokkos_def.hpp

```diff
@@ -180,6 +180,60 @@ namespace MueLu {
 
     return graph_;
   }
+  
+  template <class LocalOrdinal, class GlobalOrdinal, class DeviceType>
+  void 
+  Aggregates_kokkos<LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::ComputeNodesInAggregate(LO_view & aggPtr, LO_view & aggNodes, LO_view & unaggregated) const {
+    LO numAggs  = GetNumAggregates();
+    LO numNodes = vertex2AggId_->getLocalLength();
+    auto vertex2AggId = vertex2AggId_->getDeviceLocalView(Xpetra::Access::ReadOnly);
+    typename aggregates_sizes_type::const_type aggSizes = ComputeAggregateSizes(true);
+    LO INVALID = Teuchos::OrdinalTraits<LO>::invalid();
+
+    LO currNumUnaggregated=0;
+
+    // Construct the "rowptr" and the counter
+    // This is completely egregious to perform on host, but doesn't seem parallelizeable; is there a way to improve this?
```

Comment:
Ah shoot Github just ate my comment when I clicked away.

I think the hard part boils down to the fact that a CRS matrix has a `rowPtr`, but I don't have `aggPtr`. Instead, I have `aggSizes`, so if I parallelize on device over aggregates (which is probably the most logical choice), each aggregate needs to compute its own `aggPtr[i]=aggSizes[0]+aggSizes[1]+...+aggSizes[i]`. I feel like that would introduce some terrible thread balancing without doing some sort of smart divide and conquer strategy, but I don't know how I'd do it. Nonetheless, it could very well be faster than running on host in the first place, so it might be worthwhile.

Reply:
@GrahamBenHarper That's a scan pattern and it actually can be done in parallel. This will transform aggSizes to aggPtr (assuming aggSizes length is ``numAggs``, and aggPtr length is ``numAggs + 1``).

```
    Kokkos::parallel_scan(Kokkos::RangePolicy<Device>(0, numAggs + 1),
        KOKKOS_LAMBDA(local_ordinal_type i, local_ordinal_type& lsum, bool finalPass)
        {
          local_ordinal_type count = 0;
          if(i < numAggs)
            count = aggSizes(i);
          if(finalPass)
            aggPtr(i) = lsum;
          lsum += count;
        });
```
This version doesn't need aggPtr to be initialized beforehand. You could also do it in-place if you have the sizes start out in ``aggPtr[0...numAggs-1]``, with ``aggPtr[numAggs] == 0``. Just replace aggSizes above with aggPtr in that case.

Reply:
thanks so much @brian-kelley!